lib/ecto/adapters/postgres/connection.ex

```diff
@@ -568,22 +568,29 @@ if Code.ensure_loaded?(Postgrex) do
 
     defp create_names(prefix, sources, pos, limit) when pos < limit do
       current =
-        case elem(sources, pos) do
-          {table, schema} ->
-            name = [create_alias(table) | Integer.to_string(pos)]
-            {quote_table(prefix, table), name, schema}
-          {:fragment, _, _} ->
-            {nil, [?f | Integer.to_string(pos)], nil}
-          %Ecto.SubQuery{} ->
-            {nil, [?s | Integer.to_string(pos)], nil}
-        end
+        sources
+        |> elem(pos)
+        |> create_name(prefix, pos)
       [current | create_names(prefix, sources, pos + 1, limit)]
     end
 
     defp create_names(_prefix, _sources, pos, pos) do
       []
     end
 
+    defp create_name({table, schema}, prefix, pos),
+      do: {quote_table(prefix, table),
+           [create_alias(table) | Integer.to_string(pos)],
+           schema}
+    defp create_name({:fragment, _, _}, _prefix, pos),
+      do: {nil, [?f | Integer.to_string(pos)], nil}
+    defp create_name(%Ecto.SubQuery{}, _prefix, pos),
+      do: {nil, [?s | Integer.to_string(pos)], nil}
+    defp create_name(%FromExpr{} = from, prefix, pos),
+      do: create_name(from.source, from.prefix || prefix, pos)
+    defp create_name(%JoinExpr{} = join, prefix, pos),
+      do: create_name(join.source, join.prefix || prefix, pos)
```

Comment:
Same comments as above.

Reply:
With the `JoinExpr`s, what is the desired behavior with respect to `FromExpr`s in this case? They aren't like direct sources, nor are they like `SubQuery`s...

Unless I did something incorrect.

Reply:
The root level will always be a FromExpr or a JoinExpr. Their source will then always be a fragment, a tuple or a subquery.

Reply:
But here, there can be a `%JoinExpr{source: %FromExpr{}}`. How deep should this go?

Reply:
I would rather get the source from the from before it is put into the join and just put that source in the join at that point, but I don't understand the codebase well enough to know where that change should occur.

Reply:
@isaacsanders oh, definitely. A FromExpr should *never* be inside a JoinExpr. My suggestion: change at least one of those functions so that they are no longer recursive and it will reveal you which kind of query this is happening, then you can go and fix the code. :)

With the assertive code the code will fail, then we "just" need to fix it. :)